apps/api/src/report/report.controller.ts

```diff
@@ -171,16 +171,65 @@ export class ReportController {
     @Query() { from, to }: ReportPeriodDTO,
     @User() user: EmployeeRO,
   ): Promise<object[] | { url: string }> {
-    const data = await this.reportService.extractMilestoneData({ to, from }, user?.ha_pcn_id);
-    if (
-      data?.length > 10 &&
-      process.env.NODE_ENV !== 'test' &&
-      process.env.RUNTIME_ENV !== 'local'
-    ) {
-      const key = `ien-milestone-data-extract_${from}-${to}_${user?.user_id}_${Date.now()}`;
-      await this.reportS3Service.uploadFile(key, data);
-      return { url: await this.reportS3Service.generatePresignedUrl(key) };
+    return this.extractData(
+      { from, to },
+      user,
+      'extract-milestone',
+      'milestone',
+      this.reportService.extractMilestoneData.bind(this.reportService),
+    );
+  }
+
+  private shouldUseS3(): boolean {
+    return process.env.NODE_ENV !== 'test' && process.env.RUNTIME_ENV !== 'local';
+  }
+
+  private generateS3Key(
+    from: string,
+    to: string,
+    userId: string | undefined | null,
```

Comment:
Nit - It seems strange that user Id could be both undefined or null. 
If the user id is undefined, shouldn't we throw an error?

Reply:
Thanks, good catch! fixed.